bastionzero/policy/attributes.go

```diff
@@ -0,0 +1,307 @@
+// policy model
+type PolicyModelInterface interface {
+	// SetID sets the policy model's ID attribute.
+	SetID(value types.String)
+	// SetName sets the policy model's name attribute.
+	SetName(value types.String)
+	// SetType sets the policy model's type attribute.
+	SetType(value types.String)
+	// SetDescription sets the policy model's description attribute.
+	SetDescription(value types.String)
+	// SetSubjects sets the policy model's subjects attribute.
+	SetSubjects(value types.Set)
+	// SetGroups sets the policy model's groups attribute.
+	SetGroups(value types.Set)
+
+	// GetSubjects gets the policy model's subjects attribute.
+	GetSubjects() types.Set
+	// GetGroups gets the policy model's groups attribute.
+	GetGroups() types.Set
+}
+
+// SetBasePolicyAttributes populates base policy attributes in the TF schema
+// from a base policy
+func SetBasePolicyAttributes(ctx context.Context, schema PolicyModelInterface, basePolicy policies.PolicyInterface, modelIsDataSource bool) {
+	schema.SetID(types.StringValue(basePolicy.GetID()))
+	schema.SetName(types.StringValue(basePolicy.GetName()))
+	schema.SetType(types.StringValue(string(basePolicy.GetPolicyType())))
+	schema.SetDescription(types.StringValue(basePolicy.GetDescription()))
+
+	// Preserve null in schema if refreshed list is empty list.
+	//
+	// If we don't include this logic, then we will get "Provider produced
+	// inconsistent result after apply" error when user sets null value in
+	// config because Flatten() returns an empty set if slice is empty which is
+	// not consistent.
+	//
+	// Additionally, we always set the value in the schema if the model is a
+	// data source because it's easier to work with empty valued, computed
+	// collection attributes in data sources than null ones.
+	if !schema.GetSubjects().IsNull() || len(basePolicy.GetSubjects()) != 0 || modelIsDataSource {
+		schema.SetSubjects(FlattenPolicySubjects(ctx, basePolicy.GetSubjects()))
+	}
+	if !schema.GetGroups().IsNull() || len(basePolicy.GetGroups()) != 0 || modelIsDataSource {
+		schema.SetGroups(FlattenPolicyGroups(ctx, basePolicy.GetGroups()))
+	}
+}
+
+func BasePolicyResourceAttributes(policyType policytype.PolicyType) map[string]schema.Attribute {
+	return map[string]schema.Attribute{
+		"id": schema.StringAttribute{
+			Computed: true,
+			PlanModifiers: []planmodifier.String{
+				stringplanmodifier.UseStateForUnknown(),
+			},
+			Description: "The policy's unique ID.",
+		},
+		"name": schema.StringAttribute{
+			Required:    true,
+			Description: "The policy's name.",
+			Validators: []validator.String{
+				stringvalidator.LengthAtLeast(1),
+			},
+		},
+		"type": schema.StringAttribute{
+			Description: fmt.Sprintf("The policy's type (constant value `%s`).", policyType),
+			Computed:    true,
+			Default:     stringdefault.StaticString(string(policyType)),
+		},
+		"description": schema.StringAttribute{
+			Optional:    true,
+			Computed:    true,
```

Comment:
I forget which Terraform issue in the framework repo discusses why it works like this but basically we still need `Computed: true` because the way the `Default` feature works is through a plan modifier.

https://github.com/bastionzero/terraform-provider-bastionzero/blob/4c38dab3292e74e088f32121db2b4792fc1faa08/bastionzero/policy/attributes.go#L93-L100

 And you'll get an error if the plan modifier changes something without `Computed` being set to `true`.

Here is an example of how they use it: https://github.com/hashicorp/terraform-provider-scaffolding-framework/blob/63c8212f2b0843be76c2f33b928c93f16406536f/internal/provider/example_resource.go#L53-L56

Also, I'm pretty sure I added some acceptance tests that check updating the description attribute for policy, so we're covered there.

Reply:
Ok yup that clears it up...i think you mentioned this is to me before in regard to having a default value of empty string instead of null.